### PR TITLE
Preserve leaf-ness and requires_grad-ness in minified repros

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102972
* __->__ #102899

Also some minor refactoring

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @bertmaher

### CODE REVIEW

**Comment by vadimkantorov** [comment]
@ezyang Is https://github.com/pytorch/pytorch/issues/93859 related somehow? E.g. if there's a way for serializing a tensor object very precisely for dynamo repros, it might be good to expose it in general to be used for other kinds of repros / pasting directly in GitHub issues/code examples

**Comment by ezyang** [comment]
Yes it is! This is what InputWriter is intended to do. For dense tensors it is able to do striding and sharing, and I need autograd fidelity now so I am designing out how to do that

**Comment by vadimkantorov** [comment]
Please make it an API available even outside the Dynamo context 😀 Ideally there should directly be a very discoverable base64 format option (without doing base64 encode of some pickle first)

It would help eventually for posting sparse tensor repros

**Comment by vadimkantorov** [comment]
Please make it an API available even outside the Dynamo context 😀 Ideally there should directly be a very discoverable base64 format option (without doing base64 encode of some pickle first)

It would help eventually for posting sparse tensor repros

**Comment by ezyang** [comment]
Sure.

I'm a little more skeptical about base64 though, is megabytes of base64 goo really better than figuring out some convenient way to host binary files 🤔

**Comment by vadimkantorov** [comment]
Hmm :) for megabytes - maybe not, but for smaller repro tensors, and especially for the tensor metadata - I think it makes some sense

File: torch/_dynamo/backends/debugging.py [hunk 1]
```diff
@@ -113,3 +117,18 @@ def relu_accuracy_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs):
...
+def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs):
+    # Require at least one non-trivial thing in the graph,
+    # see https://github.com/pytorch/pytorch/issues/102898
+    for node in gm.graph.nodes:
+        if node.op == "call_function":
+            break
+    else:
+        return gm
+    for t in example_inputs:
+        if t.grad_fn is not None:
```
Comment (albanD) [inline torch/_dynamo/backends/debugging.py:132]: nit to keep in line with PR title :p 
```suggestion
        if not t.is_leaf:
```

File: torch/_dynamo/debug_utils.py [hunk 2]
```diff
@@ -291,16 +294,20 @@ def clone_inputs_retaining_gradness(example_inputs):
     return cloned_inputs
 
 
-def run_fwd_maybe_bwd(gm, args, only_fwd=False):
+def run_fwd_maybe_bwd(gm, args, only_fwd=False, disable_clone=False):
     """
     Runs a forward and possibly backward iteration for a given mod and args.
+
+    When disable_clone is True, we will use args as-is without cloning.
+    This is higher fidelity but we may destroy the args in the process.
```
Comment (albanD) [inline torch/_dynamo/debug_utils.py:302]: Is that in the case where the gm performs any inplace operation on the inputs? Or something else?
Could we clarify when they get destroyed?

**Review by albanD — APPROVED** [review]
Sounds good, just some nits

File: torch/_dynamo/debug_utils.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/debug_utils.py:302]: Yes, gm may mutate the inputs, thereby "destroying" them.

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**